python/BioSimSpace/Parameters/Protocol/_protocol.py

```diff
@@ -557,3 +572,167 @@ def _find_force_field(forcefield):
 
     # Return the force field.
     return ff
+
+def _get_disulphide_bonds(molecule, tolerance=1.2, max_distance=_Length(6, "A"),
+        property_map={}):
+    """Internal function to generate LEaP records for disulphide bonds.
+
+       Parameters
+       ----------
+
+       molecule : Sire.Mol.Molecule
+           The molecule of interest.
+
+       Returns
+       -------
+
+       bond_records : [str]
+           A list of LEaP formatted bond records.
+
+       tolerance : float
+           The tolerance to use when searching for bonds.
+
+       max_distance : :class:`Length <BioSimSpace.Types.Length>`
+           The maximum distance between atoms when searching for disulphide
+           bonds.
+
+       property_map : dict
+           A dictionary that maps system "properties" to their user defined
+           values. This allows the user to refer to properties with their
+           own naming scheme, e.g. { "charge" : "my-charge" }
+    """
+
+    if not isinstance(molecule, _SireMol.Molecule):
+        raise TypeError("'molecule' must be of type 'Sire.Mol.Molecule'")
+
+    if not isinstance(tolerance, (int, float)):
+        raise TypeError("'tolerance' must be of type 'float'")
+    tolerance = float(tolerance)
+    if tolerance < 1:
+        raise ValueError("'tolerance' must be >= 1.0.")
+
+    if not isinstance(max_distance, _Length):
+        raise ValueError("'max_distance' must be of type 'BioSimSpace.Types.Length'")
+    max_radius2 = max_distance.angstroms().value()**2
+
+    if not isinstance(property_map, dict):
+        raise ValueError("'property_map' must be of type 'dict'")
+
+    # Create a copy of the molecule.
+    mol = molecule.__deepcopy__()
+
+    # Get the connectivity of the molecule.
+    conn = _SireMol.Connectivity(mol,
+                                 _SireMol.CovalentBondHunter(tolerance, max_radius2),
+                                 property_map)
+
+    # Add this as a molecule property.
+    mol = mol.edit().setProperty("connectivity", conn).molecule().commit()
+
+    # Create the search query.
+    query =_SireMol.Select("bonds from element S to element S")
+
+    # Try searching for disulphide bonds.
+    try:
+        disulphides = query(mol, property_map)
+    except:
+        disulphides = []
+
+    # Create a list to store the LEaP bond record strings.
+    bond_records = []
+
+    # Loop over the disulphide bonds and generate the records.
+    for bond in disulphides:
```

Comment:
Could this loop be made simpler with the new sire API?

Reply:
Sorry, in what way were you thinking? It uses search functionality from 2023 to search for the bonds, i.e.:

```python
# Create the search query.
query =_SireMol.Select("bonds from element S to element S")
```

I need to use the legacy API so that the property map can be respected during the search, i.e. why I use `Sire.Mol.Select` directly.

Were you thinking that the body of the loop itself could be simplified?

Reply:
Yes, I was under the impression there might be some new `sire` 2023 syntax to simplify the loop, but I don't know much about it so might be wrong.

Reply:
Yes - there is 'apply' that can be used to apply a function to each result from a search (each item in a container). There may also be ways to get the names and residues via single calls. I don't think this is needed, but I think it could be a useful exercise for me to look at how this could be simplified.

Reply:
In this case there will likely only be a handful of bonds returned by the search so I don't think anything will speed things up much, if it all. That said, it would be good to know what's possible with the new API, and where it will really benefit. The current code seems pretty easy to follow as is. (It's only a few lines, after all.)

Reply:
I agree - I don't think there is much benefit in this case. It will be a useful exercise to try to apply the new api to a real world example :-)

Reply:
I was thinking that the restraint generation for ABFE might be able to be handled natively with the new trajectory analysis and searching code. It might be good to take a look at the current implementation with you at some point to see.